Orange/widgets/visualize/owlinearprojection.py

```diff
@@ -1,181 +1,252 @@
+            controlWidth=50, alignment=Qt.AlignRight,
+            callback=self._n_attrs_changed)
+        gui.rubber(box)
 
-    def paint(self, painter, option, widget):
-        pass
+        self.last_run_n_attrs = None
 
-    def boundingRect(self):
-        return QRectF()
+    def before_running(self):
+        """
+        Disable the spin for number of attributes before running and
+        enable afterwards. Also, if the number of attributes is different than
+        in the last run, reset the saved state (if it was paused).
+        """
+        if self.n_attrs != self.last_run_n_attrs:
+            self.saved_state = None
+            self.saved_progress = 0
+        if self.saved_state is None:
+            self.scores = []
+            self.rank_model.clear()
+        self.last_run_n_attrs = self.n_attrs
 
-    def viewTransformChanged(self):
-        self.__updateLabelPos()
+    def check_preconditions(self):
+        if not super().check_preconditions():
+            return False
+        if not self.master.data:
+            return False
+        if not self.master.data.domain.class_var:
+            return False
+        self.n_attrs_spin.setMaximum(self.master.n_cont_var)
+        return True
+
+    def state_count(self):
+        n_all_attrs = len(self.attrs)
+        n_attrs = self.n_attrs
+        return ceil(0.5 * factorial(n_all_attrs) / (factorial(n_all_attrs - n_attrs) * n_attrs))
+
+    def iterate_states(self, state):
+        if state is None:  # on the first call, compute order
+            attrs = self._score_heuristic()
+            self.attrs = attrs
+        for c in combinations(attrs, self.n_attrs):
+            for p in islice(permutations(c[1:]), ceil(factorial(len(c) - 1)/2)):
+                yield (c[0], ) + p
+
+    def compute_score(self, state):
+        master = self.master
+        variables = [s for s in state]
+        master.prepare_plot_data(variables)
+        X = master.plotdata.embedding_coords
+        Y = master.data.Y
+        if X.shape[0] < self.minK:
+            return
+        n_neighbors = min(self.minK, len(X) - 1)
+        knn = NearestNeighbors(n_neighbors=n_neighbors).fit(X)
+        ind = knn.kneighbors(return_distance=False)
+        if master.data.domain.has_discrete_class:
+            return -np.sum(Y[ind] == Y.reshape(-1, 1)) / n_neighbors / len(Y)
+        else:
+            return -r2_score(Y, np.mean(Y[ind], axis=1)) * \
+                   (len(Y) / len(master.data))
+
+    def _score_heuristic(self):
+        def normalized(col):
+            if not col.size:
+                return col.copy()
+            col_min, col_max = np.nanmin(col), np.nanmax(col)
+            if np.isnan(col_min):
+                return col.copy()
+            span = col_max - col_min
+            return (col - col_min) / (span or 1)
+        mdomain = self.master.data.domain
+        keys = [i for i, a in enumerate(mdomain.attributes) if a.is_continuous]
```

Comment:
Vizrank must process whatever the widget can show. If it can show discrete variables, they must be ranked. If it can't, let the widget provide a domain without them.